compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala

```diff
@@ -649,29 +649,334 @@ abstract class GenJSCode extends plugins.PluginComponent
-            "repeated parameters.")
+            "Scala.js-defined JS classes cannot have default parameters.")
       }
 
-      // Implementation restriction
-      for (tree <- secondaryCtorTrees) {
-        reporter.error(tree.pos,
-            "Implementation restriction: Scala.js-defined JS classes cannot " +
-            "have secondary constructors")
+      withNewLocalNameScope {
+        val ctors: List[js.MethodDef] = constructorTrees.flatMap { tree =>
+          genMethodWithCurrentLocalNameScope(tree)
+        }
+
+        val dispatch =
+          genJSConstructorExport(constructorTrees.map(_.symbol))
+        val js.MethodDef(_, dispatchName, dispatchArgs, dispatchResultType,
+            dispatchResolution) = dispatch
+
+        val isConstructor = ctors.map(_.name.name).toSet
+
+        val constructorTree = mkConstructorTree(ctors, isConstructor)
+
+        val overloadIdent = freshLocalIdent("overload")
+
+        // Section containing the overload resolution and casts of parameters
+        val overloadSelection = mkOverloadSelection(constructorTree,
+            overloadIdent, dispatchResolution, isConstructor)
+
+        /* Section containing all the code executed before the call to `this`
+         * for every secondary constructor.
+         */
+        val prePrimaryCtorBody =
+          constructorTree.mkPreSuperCall(overloadIdent, isConstructor)
+
+        val primaryCtorBody = constructorTree.rootBody
+
+        /* Section containing all the code executed after the call to this for
+         * every secondary constructor.
+         */
+        val postPrimaryCtorBody =
+          constructorTree.mkPostSuperCall(overloadIdent, isConstructor)
+
+        val newBody = js.Block(overloadSelection ::: prePrimaryCtorBody ::
+            primaryCtorBody :: postPrimaryCtorBody :: Nil)
+
+        js.MethodDef(static = false, dispatchName, dispatchArgs, jstpe.NoType,
+            newBody)(dispatch.optimizerHints, dispatch.hash)
+      }
+    }
+
+    private class ConstructorTree(overrideNum: Int, method: js.MethodDef,
+        subConstructors: List[ConstructorTree]) {
+
+      def methodName: String = method.name.name
+
+      def hasSubConstructors: Boolean = subConstructors.nonEmpty
+
+      lazy val overrideNumBounds: (Int, Int) =
+        if (subConstructors.isEmpty) (overrideNum, overrideNum)
+        else (subConstructors.head.overrideNumBounds._1, overrideNum)
+
+      def get(methodName: String): Option[ConstructorTree] = {
+        if (methodName == this.methodName) {
+          Some(this)
+        } else {
+          subConstructors.iterator.map(_.get(methodName)).collectFirst {
+            case Some(node) => node
+          }
+        }
+      }
+
+      def getOverrideNum(methodName: String): Int =
+        get(methodName).fold(-1)(_.getOverrideNum)
+
+      def rootBody: js.Tree = method.body
+
+      private def getOverrideNum: Int = overrideNum
```

Comment:
Avoid overloading with one variant taking a list of parameters and the other having no `()`. This is useless anyway, as one can use `overrideNum` directly.